tpu-client/src/nonblocking/tpu_client.rs

```diff
@@ -847,6 +826,46 @@ impl LeaderTpuService {
                 }
             }
         }
+
+        Ok(())
+    }
+
+    async fn run_slot_watcher(
+        recent_slots: RecentLeaderSlots,
+        pubsub_client: Option<PubsubClient>,
+        exit: Arc<AtomicBool>,
+    ) -> Result<()> {
+        let Some(pubsub_client) = pubsub_client else {
+            return Ok(());
+        };
+
+        let (mut notifications, unsubscribe) = pubsub_client.slot_updates_subscribe().await?;
+
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                break;
+            }
+
+            while let Ok(Some(update)) =
+                timeout(Duration::from_millis(10), notifications.next()).await
+            {
+                let current_slot = match update {
+                    // This update indicates that a full slot was received by the connected
+                    // node so we can stop sending transactions to the leader for that slot
+                    SlotUpdate::Completed { slot, .. } => slot.saturating_add(1),
+                    // This update indicates that we have just received the first shred from
+                    // the leader for this slot and they are probably still accepting transactions.
+                    SlotUpdate::FirstShredReceived { slot, .. } => slot,
+                    _ => continue,
+                };
+                recent_slots.record_slot(current_slot);
+            }
+        }
+
+        drop(notifications);
```

Comment:
why do you want to drop them explicitly?

Reply:
Might be nice to keep the comment:
```
// `notifications` requires a valid reference to `pubsub_client`
// so `notifications` must be dropped before moving `pubsub_client`
```